src/frontend/SignatureMismatch.ml

```diff
@@ -170,34 +179,56 @@ let promote es promotions =
       let open UnsizedType in
       let emeta = exp.emeta in
       match prom with
-      | RealPromotion when is_int_type emeta.type_ ->
+      | IntToRealPromotion when is_int_type emeta.type_ ->
           Ast.
             { expr= Ast.Promotion (exp, UReal, emeta.ad_level)
             ; emeta= {emeta with type_= promote_array emeta.type_ UReal} }
-      | ComplexPromotion when not (is_complex_type emeta.type_) ->
+      | (IntToComplexPromotion | RealToComplexPromotion)
+        when not (is_complex_type emeta.type_) ->
           { expr= Promotion (exp, UComplex, emeta.ad_level)
           ; emeta= {emeta with type_= promote_array emeta.type_ UComplex} }
       | _ -> exp )
 
-let returntype env name args =
+let promotion_cost p =
+  match p with
+  | None -> 0
+  | RealToComplexPromotion | IntToRealPromotion -> 1
+  | IntToComplexPromotion -> 2
+
+let matching_function env name args =
   (* NB: Variadic arguments are special-cased in the typechecker and not handled here *)
   let name = Utils.stdlib_distribution_name name in
-  Environment.find env name
-  |> List.filter_map ~f:extract_function_types
-  |> List.sort ~compare:(fun (x, _, _, _) (y, _, _, _) ->
-         UnsizedType.compare_returntype x y )
-  (* Check the least return type first in case there are multiple options (due to implicit UInt-UReal conversion), where UInt<UReal *)
-  |> List.fold_until ~init:[]
-       ~f:(fun errors (rt, tys, funkind_constructor, _) ->
-         match check_compatible_arguments 0 tys args with
-         | Ok p -> Stop (Ok (rt, funkind_constructor, p))
-         | Error e -> Continue (((rt, tys), e) :: errors) )
-       ~finish:(fun errors ->
-         let errors =
-           List.sort errors ~compare:(fun (_, e1) (_, e2) ->
-               compare_errors e1 e2 ) in
-         let errors, omitted = List.split_n errors max_n_errors in
-         Error (errors, not (List.is_empty omitted)) )
+  let unique_minimum_promotion ps =
+    let size (_, _, p) =
+      List.fold ~init:0 ~f:(fun acc p -> acc + promotion_cost p) p in
+    let ps =
+      List.sort ~compare:(fun p1 p2 -> Int.compare (size p1) (size p2)) ps in
+    match ps with
+    | (((rt, tys), _, p1) as ans1) :: ((_, _, p2) as ans2) :: _
+      when size ans1 = size ans2 ->
+        Error (Some ((rt, tys), PromotionConflict (p1, p2)))
```

Comment:
Instead of sorting and checking the first two items let's compute the minimum and then select all the items equal to the minimum.